From 168682fdf2f5579e3ee002ed4c0398fb9ff6f7cf Mon Sep 17 00:00:00 2001
From: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Date: Wed, 31 Dec 2025 11:09:55 +0800
Subject: [PATCH 13/14] plat: imx94: add SIP cpu_off service

Added SIP cpu_off service for imx94.
Enabled workaround HANDLE_NS_EA_EL3_IRQ_WA.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Upstream-Status: Pending
---
 plat/imx/common/imx_sip_handler.c          | 4 ++--
 plat/imx/common/imx_sip_svc.c              | 2 +-
 plat/imx/common/include/imx_sip_svc.h      | 2 +-
 plat/imx/imx9/imx94/imx94_bl31_setup.c     | 7 +++++++
 plat/imx/imx9/imx94/include/platform_def.h | 3 +++
 plat/imx/imx9/imx94/platform.mk            | 5 +++++
 6 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/plat/imx/common/imx_sip_handler.c b/plat/imx/common/imx_sip_handler.c
index 2788b8aad..a9e9a5705 100644
--- a/plat/imx/common/imx_sip_handler.c
+++ b/plat/imx/common/imx_sip_handler.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019,2024 NXP
+ * Copyright 2019,2024-2025 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -391,7 +391,7 @@ int imx_bbsm_handler(uint32_t smc_fid, u_register_t x1, void *handle)
 }
 #endif
 
-#if defined(PLAT_imx93) || defined(PLAT_imx95) || defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp)
+#if defined(PLAT_imx93) || defined(PLAT_imx95) || defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp) || defined(PLAT_imx94)
 uint64_t imx_sip_cpuoff_handler(uint32_t smc_fid,
 			u_register_t x1,
 			u_register_t x2,
diff --git a/plat/imx/common/imx_sip_svc.c b/plat/imx/common/imx_sip_svc.c
index c9242c745..8be50a372 100644
--- a/plat/imx/common/imx_sip_svc.c
+++ b/plat/imx/common/imx_sip_svc.c
@@ -127,7 +127,7 @@ static uintptr_t imx_sip_handler(unsigned int smc_fid,
 	case IMX_SIP_GET_PARTITION_NUMBER:
 		return imx_get_partition_number(handle);
 #endif
-#if defined(PLAT_imx93) || defined(PLAT_imx95) || defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp)
+#if defined(PLAT_imx93) || defined(PLAT_imx95) || defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp) || defined(PLAT_imx94)
 	case IMX_SIP_CPU_OFF:
 		SMC_RET1(handle, imx_sip_cpuoff_handler(smc_fid, x1, x2, x3, x4));
 #endif
diff --git a/plat/imx/common/include/imx_sip_svc.h b/plat/imx/common/include/imx_sip_svc.h
index 6c9422066..a9d8bb129 100644
--- a/plat/imx/common/include/imx_sip_svc.h
+++ b/plat/imx/common/include/imx_sip_svc.h
@@ -179,7 +179,7 @@ int imx_src_handler(uint32_t smc_fid, u_register_t x1,
 		    u_register_t x2, u_register_t x3,
 		    u_register_t x4, void *handle);
 #endif
-#if defined(PLAT_imx93) || defined(PLAT_imx95) || defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp)
+#if defined(PLAT_imx93) || defined(PLAT_imx95) || defined(PLAT_imx8mm) || defined(PLAT_imx8mn) || defined(PLAT_imx8mp) || defined(PLAT_imx94)
 uint64_t imx_sip_cpuoff_handler(uint32_t smc_fid, u_register_t x1,
     u_register_t x2, u_register_t x3, u_register_t x4);
 #endif
diff --git a/plat/imx/imx9/imx94/imx94_bl31_setup.c b/plat/imx/imx9/imx94/imx94_bl31_setup.c
index 5c35da6b4..13d1ab053 100644
--- a/plat/imx/imx9/imx94/imx94_bl31_setup.c
+++ b/plat/imx/imx9/imx94/imx94_bl31_setup.c
@@ -22,6 +22,9 @@
 #include "../drivers/arm/gic/v3/gicv3_private.h"
 
 #include <ele_api.h>
+#include <imx_cpu_off.h>
+#include <imx_interrupt_mgmt.h>
+#include <imx_intercore_svc.h>
 #include <imx8_lpuart.h>
 #include <imx_scmi_client.h>
 #include <platform_def.h>
@@ -165,6 +168,10 @@ void bl31_platform_setup(void)
 	/* get soc info */
 	ele_get_soc_info();
 
+	imx_el3_interrupt_config();
+	imx_intercore_svc_init();
+	imx_cpuoff_init();
+
 	plat_imx9_scmi_setup();
 }
 
diff --git a/plat/imx/imx9/imx94/include/platform_def.h b/plat/imx/imx9/imx94/include/platform_def.h
index fb463ed28..78f44002c 100644
--- a/plat/imx/imx9/imx94/include/platform_def.h
+++ b/plat/imx/imx9/imx94/include/platform_def.h
@@ -93,4 +93,7 @@
 #define IMX_TRUSTY_STACK_SIZE 0x200
 #define TRUSTY_SHARED_MEMORY_OBJ_SIZE (12 * 1024)
 
+/*Use SGI-8 as Inter-Core Service IRQ */
+#define INTERCORE_SVC_IRQ   0x8
+
 #endif /* platform_def.h */
diff --git a/plat/imx/imx9/imx94/platform.mk b/plat/imx/imx9/imx94/platform.mk
index 84afa04b9..ad2448308 100644
--- a/plat/imx/imx9/imx94/platform.mk
+++ b/plat/imx/imx9/imx94/platform.mk
@@ -41,6 +41,9 @@ BL31_SOURCES		+=	drivers/arm/css/scmi/scmi_common.c		\
 				lib/cpus/aarch64/cortex_a55.S			\
 				drivers/delay_timer/delay_timer.c		\
 				drivers/delay_timer/generic_delay_timer.c	\
+				plat/imx/common/imx_cpu_off.c			\
+				plat/imx/common/imx_interrupt_mgmt.c		\
+				plat/imx/common/imx_intercore_svc.c		\
 				plat/imx/common/imx_sip_handler.c		\
 				plat/imx/common/imx_sip_svc.c			\
 				plat/imx/common/ele_api.c			\
@@ -60,6 +63,8 @@ COLD_BOOT_SINGLE_CPU := 1
 ERRATA_A55_1530923 := 1
 IMX_CRRM := 0
 
+HANDLE_NS_EA_EL3_IRQ_WA	:=	1
+
 BL32_BASE               ?=      0x8C000000
 BL32_SIZE               ?=      0x02000000
 $(eval $(call add_define,BL32_BASE))
-- 
2.43.0

